packages/examples/cvat/exchange-oracle/src/handlers/completed_escrows.py

```diff
@@ -30,382 +27,259 @@
-        logger = self.logger
+            job_assignment = job.latest_assignment
+            job_annotations[job.cvat_id] = FileDescriptor(
+                filename="project_{}-task_{}-job_{}-user_{}-assignment_{}.zip".format(
+                    job.cvat_project_id,
+                    job.cvat_task_id,
+                    job.cvat_id,
+                    job_assignment.user.cvat_id,
+                    job_assignment.id,
+                ),
+                file=job_annotations_file,
+            )
+    return job_annotations
+
+
+def _handle_escrow_validation(
+    logger: logging.Logger,
+    escrow_address: str,
+    chain_id: int,
+) -> bool:
+    try:
+        validate_escrow(chain_id, escrow_address)
+    except Exception as e:
+        logger.exception(f"Failed to handle completed projects for escrow {escrow_address}: {e}")
+        return False
+
+    # Need to work in separate transactions for each escrow, as a failing DB call
+    # (e.g. a failed lock attempt) will abort the transaction. A nested transaction
+    # can also be used for handling this.
+    with SessionLocal.begin() as session:
+        try:
+            escrow_projects = cvat_service.get_projects_by_escrow_address(
+                session, escrow_address, limit=None, for_update=ForUpdateParams(nowait=True)
+            )
+        except sa_errors.OperationalError as ex:
+            if isinstance(ex.orig, db_errors.LockNotAvailable):
+                return False
+            raise
+
+        _export_escrow_annotations(logger, chain_id, escrow_address, escrow_projects, session)
+
+    return True
 
-        # Here we can have several projects per escrow, so the handling is done in project groups
-        # Find potentially finished escrows first
+
+def handle_escrows_validations(logger: logging.Logger) -> None:
+    with SessionLocal.begin() as session:
+        escrow_validations = cvat_service.prepare_escrows_for_validation(
+            session,
+            limit=CronConfig.track_completed_escrows_chunk_size,
+        )
+
+    unhandled_validations = set(validation_id for validation_id, _, __ in escrow_validations)
+    handled_validations = set()
+    try:
+        for validation_id, escrow_address, chain_id in escrow_validations:
+            if _handle_escrow_validation(logger, escrow_address, chain_id):
+                handled_validations.add(validation_id)
+    finally:
         with SessionLocal.begin() as session:
-            escrows_with_completed_projects = cvat_service.get_escrows_by_project_status(
+            cvat_service.update_escrow_validation_statuses_by_ids(
                 session,
-                ProjectStatuses.completed,
-                included_types=[TaskTypes.image_skeletons_from_boxes],
-                limit=CronConfig.track_completed_escrows_chunk_size,
+                unhandled_validations - handled_validations,
+                EscrowValidationStatuses.awaiting,
             )
```

Comment:
I don't think this request optimization makes sense here. Probably, we need to do something similar to webhook processing here, so 1 by 1. Escrow export is a lengthy operation that takes about 5 minutes, so we're not going to launch many of them in one status update (cron) call.

Reply:
Makes sense, thank you. I've changed it to per-validation update and decreased the default limit.